File: GeeksforGeeks_Solutions/C++/October_06.cpp
#include <iostream>
using namespace std;

class Solution
{
public:
    void solve(vector<vector<char>> &grid, int i, int j)
    {

        if (i < 0 || i >= grid.size() || j < 0 || j >= grid[0].size() || grid[i][j] == '0')
            return;
        vector<int> dx = {-1, 1, 0, 0, -1, 1, -1, 1};
        vector<int> dy = {0, 0, -1, 1, 1, 1, -1, -1};
        grid[i][j] = '0';
        for (int k = 0; k < 8; k++)
        {
            int new_x = i + dx[k];
            int new_y = j + dy[k];
            solve(grid, new_x, new_y);
        }
        return;
    }
    int numIslands(vector<vector<char>> &grid)
    {
        int counter = 0;
        for (int i = 0; i < grid.size(); i++)
        {
            for (int j = 0; j < grid[0].size(); j++)
            {
                if (grid[i][j] == '1')
                {
                    solve(grid, i, j);
                    counter++;
                }
            }
        }
        return counter;
    }
};

File: GeeksforGeeks_Solutions/python/October_06.py
class Solution:
    def numIslands(self, grid):
        if not grid:
            return 0

        n, m = len(grid), len(grid[0])
        count = 0

        def dfs(x, y):
            if x < 0 or x >= n or y < 0 or y >= m or grid[x][y] == 0:
                return
            grid[x][y] = 0
            for dx in [-1, 0, 1]:
                for dy in [-1, 0, 1]:
                    if not (dx == 0 and dy == 0):
                        dfs(x + dx, y + dy)

        for i in range(n):
            for j in range(m):
                if grid[i][j] == 1:
                    count += 1
                    dfs(i, j)

        return count
